### PR TITLE
Add Reconnect Jitter

### PR DESCRIPTION
**Pull request by ColinSullivan1**
Add some jitter to the reconnect wait.

See:
 - https://github.com/nats-io/nats.go/issues/563
 - https://github.com/nats-io/nats.go/pull/564

Signed-off-by: Colin Sullivan <colin@synadia.com>

### CODE REVIEW

**Review by kozlovic — APPROVED** [review]
LGTM about the logic, will let others judge on the .NET language specifics.

File: src/NATS.Client/Conn.cs [hunk 1]
```diff
@@ -1526,6 +1531,15 @@ private void scheduleConnEvent(EventHandler<ConnEventArgs> connEvent, Exception
             }
         }
 
+        private void DefaultReconnectDelayHandler(object o, ReconnectDelayEventArgs args)
+        {
+            Random rand = new Random();
```
Comment (watfordgnf) [inline src/NATS.Client/Conn.cs]: This should be a static instance.

File: src/NATS.Client/Conn.cs [hunk 2]
```diff
@@ -1562,43 +1576,43 @@ private void doReconnect()
 
                 scheduleConnEvent(Opts.DisconnectedEventHandler, errorForHandler);
 
-                // TODO:  Look at using a predicate delegate in the server pool to
-                // pass a method to, but locking is complex and would need to be
-                // reworked.
                 Srv cur;
+                int wlf = 0;
```
Comment (watfordgnf) [inline src/NATS.Client/Conn.cs:1577]: `wlf`?

File: src/NATS.Client/NATS.cs [hunk 3]
```diff
@@ -186,6 +196,29 @@ internal ConnEventArgs(Connection c, Exception error = null)
...
+    {
+        internal ReconnectDelayEventArgs(int attempts)
+        {
+            Attempts = attempts;
+        }
+
+        /// <Summary>
+        /// Gets the number of times the client has traversed the
+        /// server list in attempting to reconnect.
+        /// </Summary>>
```
Comment (watfordgnf) [inline src/NATS.Client/NATS.cs]:
```suggestion
        /// </Summary>
```

File: src/NATS.Client/NATS.cs [hunk 4]
```diff
@@ -186,6 +196,29 @@ internal ConnEventArgs(Connection c, Exception error = null)
         public Exception Error { get; }
     }
 
+    /// <summary>
+    /// Provides details for the ReconnectDelayEvent.
+    /// </summary>
+    /// <remarks>
+    /// This event handler is a good place to apply backoff logic.  The associated
+    /// connection will be RECONNECTING so access or calling methods will result
```
Comment (watfordgnf) [inline src/NATS.Client/NATS.cs]:
```suggestion
    /// connection will be RECONNECTING so accessing or calling IConnection methods will result
```

File: src/NATS.Client/NATS.cs [hunk 5]
```diff
@@ -186,6 +196,29 @@ internal ConnEventArgs(Connection c, Exception error = null)
...
+    /// in undefined behavior.
```
Comment (watfordgnf) [inline src/NATS.Client/NATS.cs]:
```suggestion
    /// in undefined behavior (including deadlocks).
```

File: src/NATS.Client/NATS.cs [hunk 6]
```diff
@@ -186,6 +196,29 @@ internal ConnEventArgs(Connection c, Exception error = null)
         public Exception Error { get; }
     }
 
+    /// <summary>
+    /// Provides details for the ReconnectDelayEvent.
+    /// </summary>
+    /// <remarks>
+    /// This event handler is a good place to apply backoff logic.  The associated
```
Comment (watfordgnf) [inline src/NATS.Client/NATS.cs:203]: We may want to mention that assigning a non-default value to this event handler means the user is responsible for reconnecting backoff.

File: src/NATS.Client/Conn.cs [hunk 7]
```diff
@@ -147,9 +147,10 @@ public Options Opts
         internal Exception lastEx;
 
         Timer               ptmr = null;
-
         int                 pout = 0;
 
+        internal AutoResetEvent ReconnectDelayARE = new AutoResetEvent(false);
```
Comment (watfordgnf) [inline src/NATS.Client/Conn.cs]: This should get disposed in the `Dispose` method.

**Review by watfordgnf — COMMENTED** [review]
Some questions and comments.

File: src/NATS.Client/Conn.cs (hunk 7 above)
Comment (watfordgnf) [reply src/NATS.Client/Conn.cs]: Also, why an `ARE` instead of just `Thread.Sleep`?

File: src/Tests/IntegrationTests/TestReconnect.cs [hunk 8]
```diff
@@ -568,6 +558,92 @@ public void TestReconnectBufferBoundary()
...
+                    // Wait a bit for the reconnect loop to go into wait mode.
+                    Thread.Sleep(250);
+
+                    // Close the connection to break waiting
+                    c.Close();
+                    sw.Stop();
+
+                    // It should be around 400 ms max, but only need to check that
+                    // it's less than 30000.
+                    Assert.True(sw.ElapsedMilliseconds < 2000);
```
Comment (watfordgnf) [inline src/Tests/IntegrationTests/TestReconnect.cs]: 30 seconds or 2 seconds?

File: src/NATS.Client/Conn.cs (hunk 1 above)
Comment (ColinSullivan1) [reply src/NATS.Client/Conn.cs]: Will fix - thanks!

File: src/NATS.Client/Conn.cs (hunk 2 above)
Comment (ColinSullivan1) [reply src/NATS.Client/Conn.cs:1577]: Copied from go for consistency / portability.  It stands for "whole list failed".

File: src/Tests/IntegrationTests/TestReconnect.cs (hunk 8 above)
Comment (ColinSullivan1) [reply src/Tests/IntegrationTests/TestReconnect.cs]: Will fix...

File: src/NATS.Client/Conn.cs (hunk 7 above)
Comment (ColinSullivan1) [reply src/NATS.Client/Conn.cs]: It seemed cleaner in that author intent is expressed, but I'll test with a Sleep to see if it gets interrupted.

File: src/NATS.Client/NATS.cs (hunk 6 above)
Comment (ColinSullivan1) [reply src/NATS.Client/NATS.cs:203]: Will do, thanks!

File: src/NATS.Client/Conn.cs (hunk 7 above)
Comment (ColinSullivan1) [reply src/NATS.Client/Conn.cs]: Thread.Sleep() worked....  Simpler and easier.  Thanks!

**Comment by ColinSullivan1** [comment]
Works locally, but need to fix some flappers in CI.

**Pull request merged**